### ISSUE TITLE
toggling between `JSDoc` and `TypeScript` examples in docs

### ISSUE THREAD

**Issue by ivanhofer** [sveltejs/kit#4452]
### Describe the problem

As someone who prefers to write TypeScript code, having all examples on the docs page annotated with JSDoc makes e.g copy & pasting annoying because I have to convert it manually to TypeScript syntax. For me, a person that knows both syntaxes well, this might be a small problem.

But I already saw a few people struggle because they were new to JSDoc comments and didn't know how the equivalent should look in TypeScript code. Having a proper TypeScript example would help them a lot.

### Describe the proposed solution

A toggle could be provided somewhere on the docs page that allows to globally choose between `JSDoc` and `TypeScript`

Similar to the [Vue.js documentation](https://vuejs.org/guide/introduction.html) that allows to toggle between API styles.

### Alternatives considered

JSDoc only examples

### Importance

nice to have

**Comment by Mlocik97** [comment]
Yeah, I'm for having it unified, and ideally with just JSDocs, because beginners that doesn't know TS are totally lost in docs... and those who know TS have more likely no problem with JSDocs

**Comment by Mlocik97** [comment]
it already shows types at hover.

**Comment by ivanhofer** [comment]
Definitely a good first step, but I can imagine a proper TypeScript example would improve the developer experience.

**Comment by Rich-Harris** [comment]
Any thoughts on how we could automate this? It'd be a neat feature, but needing to maintain both variants would be prohibitive, I think

**Comment by ivanhofer** [comment]
Maybe annotate it via
```ts
@type('./[slug]', 'Load')
export async function load({ params, fetch, session, stuff }) {
   // ...
}
```
and write a custom plugin that translate it into the specific syntax:

_JSDoc_
```js
/** @type {import('./[slug]').Load} */
export async function load({ params, fetch, session, stuff }) {
   // ...
}
```

_TypeScript_
```ts
import Load from './[slug]'

export async function load: Load({ params, fetch, session, stuff }) {
   // ...
}
```

Probably needs some AST transformations or pretty good Regex skills ^^

**Comment by Mlocik97** [comment]
wouldn't it be better to just anotate it dirrectly in one of these two styles of typing and just convert it to second one, and if I know tsc can do that. See: https://www.npmjs.com/package/ts-to-jsdoc

**Comment by ivanhofer** [comment]
Sure, if this also works with type imports?

**Comment by Rich-Harris** [comment]
ts-to-jsdoc doesn't appear to work. I think we'd need to create our own special syntax as in https://github.com/sveltejs/kit/issues/4452#issuecomment-1085444682, which I can imagine proving quite costly over time

**Comment by ivanhofer** [comment]
I actually have kind of the same problems in one of my projects. I need to generate `TypeScript` or `JSDoc-JavaScript` code.
I'm doing this by calling some helper functions that output the code in the specific syntax or nothing if it doesn't match the configured syntax.

I write normal JavaScript code and then sprinkle in those functions where needed. Not the prettiest code but it works ^^
See an example [here](https://github.com/ivanhofer/typesafe-i18n/blob/main/packages/generator/src/files/generate-adapter-svelte.ts#L34).

An improved solution would probably be something like this where you write e.g. a function in an abstract form and
```ts
function({
   name: 'myFunction',
   params: [{ id: 'string' }],
   returns: 'string',
   body: 'return id'
)

// generated TypeScript code
const myFunction = (id: string): string=> {
   return id
}

// generated JavaScript code
/**
 * @param { string } id
 * @return string
 */
const myFunction = (id) => {
   return id
}
```
But such an approach would make it harder for new contributions because they would need to learn how to use such an abstraction.

But writing it in plain `TypeScript` and `Javascript` and having to keep two different examples in sync would also be a challenging task.

**Comment by danielniccoli** [comment]
> For me, a person that knows both syntaxes well, this might be a small problem.

To give some input from a beginner in TypeScript and SvelteKit: It is immensely difficult to both, learn TypeScript and SvelteKit, when also having to translate the examples from JavaScript and JSDoc (which is something I haven't heard before). Having all examples in the SK docs available in TypeScript would be _immensely_ helpful!

**Comment by benmccann** [comment]
Copying some comments Rich made on Discord here to better preserve them:

It absolutely needs to be auto-generated, yes. The big challenge is that we ideally want your settings to be persisted (rather than the page loading then there being a big layout shift while it switches from JS mode to TS mode), which implies we can't use prerendering. But we need prerendering, because the docs are driven by markdown files, and we don't have access to those files in the deployed app, just the code.

In short, this would be a huge undertaking that will involve a complete rethink of how our docs are created

It's an interesting case — we still want to avoid recomputing the HTML for every request. JS/TS and light/dark modes are examples of a situation where you want to prerender multiple options and choose between them on the edge. it's something that kit doesn't support, but something we maybe should support
